thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+
+\begin{figure}[H]
+	\caption{An example of a mesh rendering hazard}
+  \centering
+    \includegraphics[width=0.6\textwidth]{assets/3-rendering-hazard.png}
+\end{figure}
+
+Unfortunately, the issue has not been resolved in the current iteration. Some potential solutions include:
+\begin{itemize}
+	\item preventing face collision during solving,
+	\item detecting face interpenetration after solving, and marking which one is on top.
+\end{itemize}
+
+\subsubsection{Large web application bundle size}
+
+The frontend application uses multiple feature-rich libraries of significant file size footprint such as \tech{Three.js} or \tech{Material-UI}. The whole bundle size was $ 6.81\mathrm{MB} $ resulting in a non-ideal download and parsing times of several seconds.
+
+\medskip
+To analyze the size of all bundle components, \tech{BundleAnalyzerPlugin} for \tech{Webpack} was used. 
+
+\medskip
+
+\begin{figure}[H]
+	\caption{The output of the analysis is in the form of an interactive website that allows drilling down and identifying potential areas of improvement.}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-bundle.png}
+\end{figure}
+
+To optimize the application size several techniques have been used:
+
+\begin{description}
+	\item[tree-shaking] - dead code elimination technique
+	\item[minimizing] - unnecessary character removal
+	\item[chunking] - code splitting and on-demand chunk loading
+\end{description}
+
+The minimizing was done using the \tech{Terser} plugin. Chunk splitting and tree-shaking feature are included, but not enabled by default in \tech{Webpack}.
+
+\begin{table}[H]
+\caption{Incorporating optimizations reduced the bundle size massively. Cutting the load time by a large factor.}
+\label{bundle-size-optimizations}
+\begin{tabular}{|l|l|l|l|l|}
+\hline
+\multicolumn{1}{|c|}{} & \multicolumn{2}{l|}{Before optimizations} & \multicolumn{2}{l|}{After optimizations} \\ \hline
+                       & Size                & Gzipped            & Size                & Gzipped            \\ \hline
+Three.js               & 1.2MB               & 233.82KB           & 624.87KB            & 148.83KB           \\ \hline
+Material-UI            & 701.28KB            & 122.62KB           & 196.81KB            & 56.33KB            \\ \hline
+Application size       & 6.81MB              & 1.52MB             & 1.2MB               & 322.15KB           \\ \hline
+\multicolumn{1}{|l|}{Load time\footnotemark} & \multicolumn{2}{c|}{4.24s} & \multicolumn{2}{c|}{0.81s} \\ \hline
+\end{tabular}
+\footnotetext{Duration calculated between client-connection connection establishment and \texttt{load} DOM event being fired}
+\end{table}
+
+\subsubsection{Generating thumbnails for computed Guides}
+
+As of now, Guide thumbnails represent a flat crease pattern. Ideally, the Guide's image would contain a fully folded model. The functionality is not trivial to implement and due to other issues having a higher priority has not been worked on.
```

Comment:
Here, I miss the commas:

"The functionality is not trivial to implement and, due to other issues having a higher priority, has not been worked on."

Reply:
I don't think they are mandatory. I personally wouldn't pause before `due` when articulating this sentence.

Sth to be discussed.

Reply:
This is a drop-in sentence, there should be commas. If you leave only "The functionality is not trivial to implement and has not been worked on" it still makes perfect sense.

Reply:
I don't think `drop-in` is not what you meant. It is a nonrestrictive _subordinate_ clause. I thought `and` acted like a comma, but it turns out the nonrestrictive subordinate clauses are the only exception when a comma after _and_ can appear.